r2dbc-spi/src/main/java/io/r2dbc/spi/Result.java

```diff
@@ -49,4 +54,124 @@
      */
     <T> Publisher<T> map(BiFunction<Row, RowMetadata, ? extends T> mappingFunction);
 
+    /**
+     * Returns a filtered variant of the {@link Result} to selectively consume result segments matching {@link Predicate filter predicate}.
+     * <p>The returned {@link Result} is a potentially reduced view of the underlying {@link Result} to filter out unwanted result segments.  For example, filtering {@link Message} segments of
+     * {@link Message.Severity#ERROR error severity} from the result
+     * lets the result complete without an error.
+     *
+     * @param filter the non-interfering and stateless {@link Predicate} to apply to each element to determine if it should be included
+     * @return a filtered {@link Result}
+     * @throws IllegalArgumentException if {@code filter} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    Result filter(Predicate<Segment> filter);
+
+    /**
+     * Returns a mapping of the {@link Segment result segments} that are the results of a query against a database.  May be empty if the query did not return any segments.  A {@link Segment} can be
+     * only considered valid within a {@link Function mapping function} callback.
+     * <p>Consuming result {@link Segment segments} does not emit {@link Subscriber#onError(Throwable) error signals} from {@link Message#severity() error segments}.  Translation of error segments
+     * is subject to the {@code mappingFunction}.
+     *
+     * @param mappingFunction the {@link Function} that maps a {@link Segment} a to a value
+     * @param <T>             the type of the mapped value
+     * @return a mapping of the segments that are the results of a query against a database
+     * @throws IllegalArgumentException if {@code mappingFunction} is {@code null}
+     * @throws IllegalStateException    if the result was consumed
+     * @since XXX
+     */
+    <T> Publisher<T> flatMap(Function<Segment, ? extends Publisher<T>> mappingFunction);
+
+    /**
+     * Marker interface for a result segment.  Result segments represent the individual parts of a result from a query against a database.  Segments are typically update counts, data rows/{@code
+     * OUT} parameters, and messages of different severities.  Segments can also represent vendor-specific extensions.
+     *
+     * @since XXX
+     */
+    interface Segment {
+
+    }
+
+    /**
+     * Data segment consisting of {@link Row gettable data} and its {@link #metadata() RowMetadata}.  This segment can be backed by either a {@link Row} or any other type of gettable data such as
+     * {@code OUT} parameters.
+     *
+     * @since XXX
+     */
+    interface Data extends Segment, Row {
+
+        RowMetadata metadata();
```

Comment:
The presence of this accessor (which will have to be repeated when a `RowSegment` is distinguished from an `OutSegment` (see https://github.com/r2dbc/r2dbc-mssql/issues/199#issuecomment-823245179) seems to hint at `RowMetadata` being generally useful wherever `Row` is available. So, why not pull it up to `Row` instead? It would also simplify `<T> Publisher<T> map(BiFunction<Row, RowMetadata, ? extends T> mappingFunction)` to a more convenient `<T> Publisher<T> map(Function<Row, ? extends T> mappingFunction)`

Reply:
This would also be more consistent also with `Connection` / `ConnectionMetadata` or `ConnectionFactory` / `ConnectionFactoryMetadata` API designs...

Reply:
> Similar, `OutParameters` would define `OutParametersMetadata getMetadata()`.

Indeed, JDBC has `ParameterMetaData` as well.